R/theme-elements.R

```diff
@@ -81,55 +81,86 @@ NULL
 
 #' @export
 #' @rdname element
-element_blank <- function() {
-  structure(
-    list(),
-    class = c("element_blank", "element")
-  )
-}
+element <- S7::new_class("element", abstract = TRUE)
 
 #' @export
 #' @rdname element
-element_rect <- function(fill = NULL, colour = NULL, linewidth = NULL,
-  linetype = NULL, color = NULL, inherit.blank = FALSE, size = deprecated()) {
+element_blank <- S7::new_class("element_blank", parent = element)
+
+# All properties are listed here so they can easily be recycled in the different
+# element classes
+#' @include properties.R
+#' @include margins.R
+element_props <- list(
+  fill       = property_nullable(S7::class_character | S7::new_S3_class("GridPattern") | S7::class_logical),
+  colour     = property_nullable(S7::class_character | S7::class_logical),
+  family     = property_nullable(S7::class_character),
+  hjust      = property_nullable(S7::class_numeric),
+  vjust      = property_nullable(S7::class_numeric),
+  angle      = property_nullable(S7::class_numeric),
+  size       = property_nullable(S7::class_numeric),
+  lineheight = property_nullable(S7::class_numeric),
+  margin     = property_nullable(margin),
+  face       = property_choice(c("plain", "bold", "italic", "oblique", "bold.italic"), allow_null = TRUE),
+  linewidth  = property_nullable(S7::class_numeric),
+  linetype   = property_nullable(S7::class_numeric | S7::class_character),
+  lineend    = property_choice(c("round", "butt", "square"), allow_null = TRUE),
```

Comment:
We should probably add linejoin while we are doing the rewrite

Reply:
I'll happily add that